buffer.c

```diff
@@ -559,151 +468,160 @@ static int array_buffer_view_has_dimension(zval *object, zval *zv_offset, int ch
 
 	if (check_empty) {
 		int retval;
-		zval *value = buffer_view_offset_get(intern, offset);
-		retval = zend_is_true(value);
-		zval_ptr_dtor(&value);
+		zval value;
+			    buffer_view_offset_get(intern, offset, &value);
+		retval = zend_is_true(&value);
 		return retval;
 	}
 
 	return 1;
 }
 
-static void array_buffer_view_unset_dimension(zval *object, zval *zv_offset TSRMLS_DC)
+static void array_buffer_view_unset_dimension(zval *object, zval *zv_offset)
 {
-	buffer_view_object *intern = zend_object_store_get_object(object TSRMLS_CC);
+	buffer_view_object *intern = Z_BUFFER_VIEW_OBJ_P(object);
 
 	if (intern->std.ce->parent) {
-		zend_get_std_object_handlers()->unset_dimension(object, zv_offset TSRMLS_CC);
+		zend_get_std_object_handlers()->unset_dimension(object, zv_offset);
 		return;
 	}
 	
-	zend_throw_exception(NULL, "Cannot unset offsets in a typed array", 0 TSRMLS_CC);
+	zend_throw_exception(NULL, "Cannot unset offsets in a typed array", 0);
 }
 
-static int array_buffer_view_compare_objects(zval *obj1, zval *obj2 TSRMLS_DC)
+static int array_buffer_view_compare_objects(zval *obj1, zval *obj2)
 {
-	buffer_view_object *intern1 = zend_object_store_get_object(obj1 TSRMLS_CC);	
-	buffer_view_object *intern2 = zend_object_store_get_object(obj2 TSRMLS_CC);	
+	buffer_view_object *intern1 = Z_BUFFER_VIEW_OBJ_P(obj1);
+	buffer_view_object *intern2 = Z_BUFFER_VIEW_OBJ_P(obj2);
+	int cmp;
 
-	if (memcmp(intern1, intern2, sizeof(buffer_view_object)) == 0) {
+	if (intern1->type != intern2->type) {
+		return 1; /* not orderable */
+	}
+	if (intern1->length != intern2->length) {
+		return 1; /* not orderable */
+	}
+
+#define BUFFER_SWITCH_CASE(type) \
+		case buffer_view_##type:                                                          \
+			cmp = memcmp((void*) &intern1->buf.as_##type[intern1->offset],            \
+						(void*) &intern2->buf.as_##type[intern2->offset], \
+						sizeof(&intern1->buf.as_##type[intern1->offset]) * intern1->length); break;
+
+	switch (intern1->type) {
+		BUFFER_SWITCH_CASE(uint8)
+		BUFFER_SWITCH_CASE(int16)
+		BUFFER_SWITCH_CASE(uint16)
+		BUFFER_SWITCH_CASE(int32)
+		BUFFER_SWITCH_CASE(uint32)
+		BUFFER_SWITCH_CASE(float)
+		BUFFER_SWITCH_CASE(double)
```

Comment:
Why do we need this? These are all the same pointer.

Reply:
The pointer is the same yes, but the size to copy is expressed in the length of the type. 

I will streamline the code  more elegantly by introducing a temporary variable `type_size`.

Reply:
My bad, there was a `buffer_view_get_bytes_per_element` function just for that. Using it made the code much better.